consensus/src/quorum_store/utils.rs

```diff
@@ -234,14 +271,50 @@ impl ProofQueue {
 
     #[inline]
     fn dec_remaining(&mut self, author: &AccountAddress, num_txns: u64) {
-        self.remaining_txns -= num_txns;
+        self.remaining_txns_with_duplicates -= num_txns;
         self.remaining_proofs -= 1;
         if *author == self.my_peer_id {
             self.remaining_local_txns -= num_txns;
             self.remaining_local_proofs -= 1;
         }
     }
 
+    fn remaining_txns_without_duplicates(&self) -> u64 {
+        // All the batch keys for which batch_to_proof is not None. This is the set of unexpired and uncommitted proofs.
+        let unexpired_batch_keys = self
+            .batch_to_proof
+            .iter()
+            .filter(|(_, proof)| proof.is_some())
+            .map(|(batch_key, _)| batch_key)
+            .collect::<HashSet<_>>();
+        let mut remaining_txns = self
+            .txn_summary_to_batches
+            .iter()
+            .filter(|(_, batches)| {
+                batches
+                    .iter()
+                    .any(|batch_key| unexpired_batch_keys.contains(batch_key))
```

Comment:
This is a linear scan of all outstanding txns, and then a lookup. Maybe a bit expensive?

Reply:
On the realistic_env_max_load test, at about 8.5k TPS workload, calculating the remaining txns is taking 2-4ms.

Here's an optimization. What do you think?
  We are calculating remaining txns here only for the sake of checking if there is backpressure. When there is no backpressure from the old way of calculating remaining txns, we can just use the old calculation. When the old calculation says there is backpressure, we can then try to use the new way to calculate more accurately and check if there is really some backpressure.

Reply:
Might not help much as it optimizes away the cheaper part. I'm more concerned when the number of transactions is large. When it's small, it should take less time anyway.

2-4ms is probably ok.